File: Serene/Serene.Script/Common/Helper/GridEditorBase.cs
namespace Serene.Common
{
    using jQueryApi;
    using Serenity;
    using System.Runtime.CompilerServices;

    [Imported, IncludeGenericArguments(false), ScriptName("GridEditorBase")]
    public abstract class GridEditorBase<TEntity> : EntityGrid<TEntity, object>
        where TEntity: class, new()
    {
        public GridEditorBase(jQueryObject container)
            : base(container)
        {
        }
    }
}

/* 
This class has been ported to TypeScript. See GridEditorBase.ts
Code below is only a reference for those who want to use Saltaralle

namespace Serene.Common
{
    using jQueryApi;
    using Serenity;
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Runtime.CompilerServices;

    [Element("<div/>"), Editor, IdProperty("__id"), IncludeGenericArguments(false)]
    public abstract class GridEditorBase<TEntity> : EntityGrid<TEntity>, ISetEditValue, IGetEditValue
        where TEntity : class, new()
    {
        private int nextId = 1;

        public GridEditorBase(jQueryObject container)
            : base(container)
        {
        }

        protected Int32? ID(TEntity entity)
        {
            return entity.As<dynamic>().__id;
        }

        protected virtual void Save(ServiceCallOptions opt, Action<ServiceResponse> callback)
        {
            var request = opt.Request.As<SaveRequest<TEntity>>();
            var row = Q.DeepClone(request.Entity);
            int? id = row.As<dynamic>().__id;
            if (id == null)
                row.As<dynamic>().__id = nextId++;

            if (!ValidateEntity(row, id))
                return;

            var items = view.GetItems().Clone();

            if (id == null)
                items.Add(row);
            else
            {
                var index = items.IndexOf(x => ID(x) == id.Value);
                items[index] = Q.DeepExtend<TEntity>(new object().As<TEntity>(), items[index], row);
            }

            SetEntities(items);
            callback(new ServiceResponse());
        }

        protected virtual bool DeleteEntity(Int32 id)
        {
            view.DeleteItem(id);
            return true;
        }

        protected virtual bool ValidateEntity(TEntity row, Int32? id)
        {
            return true;
        }

        protected virtual void SetEntities(List<TEntity> items)
        {
            view.SetItems(items, true);
        }

        protected virtual TEntity GetNewEntity()
        {
            return new object().As<TEntity>();
        }

        protected override List<ToolButton> GetButtons()
        {
            return new List<ToolButton>
            {
                new ToolButton
                {
                    Title = GetAddButtonCaption(),
                    CssClass = "add-button",
                    OnClick = delegate
                    {
                        CreateEntityDialog(GetItemType(), dlg =>
                        {
                            var dialog = (GridEditorDialog<TEntity>)dlg;
                            dialog.OnSave = Save;
                            dialog.LoadEntityAndOpenDialog(GetNewEntity());
                        });
                    }
                }
            };
        }

        protected override void EditItem(object entityOrId)
        {
            var id = entityOrId.ToInt32().Value;
            var item = view.GetItemById(id);
            CreateEntityDialog(GetItemType(), dlg =>
            {
                var dialog = (GridEditorDialog<TEntity>)dlg;

                dialog.OnDelete = (opt, callback) => {
                    if (!DeleteEntity(id))
                        return;

                    callback(new DeleteResponse());
                };

                dialog.OnSave = Save;
                dialog.LoadEntityAndOpenDialog(item);
            });
        }

        public void GetEditValue(PropertyItem property, dynamic target)
        {
            target[property.Name] = Value;
        }

        public void SetEditValue(dynamic source, PropertyItem property)
        {
            Value = source[property.Name];
        }

        public List<TEntity> Value
        {
            get
            {
                return view.GetItems().Select(x => {
                    var y = Q.DeepClone(x);
                    Script.Delete(y, "__id");
                    return y;
                }).ToList();
            }
            set
            {
                view.SetItems((value ?? new List<TEntity>()).Select(x => {
                    var y = Q.DeepClone(x);
                    y.As<dynamic>().__id = nextId++;
                    return y;
                }).ToList(), true);
            }
        }

        protected override bool GetGridCanLoad() { return false; }
        protected override bool UsePager() { return false; }
        protected override string GetInitialTitle() { return null; }
        protected override void CreateQuickSearchInput() {}
    }
}
*/

File: Serene/Serene.Web/Modules/Common/Helpers/GridEditorBase.ts
﻿namespace Serene.Common {

    @Serenity.Decorators.registerClass([Serenity.IGetEditValue, Serenity.ISetEditValue])
    @Serenity.Decorators.editor()
    @Serenity.Decorators.element("<div/>")
    export class GridEditorBase<TEntity> extends Serenity.EntityGrid<TEntity, any>
        implements Serenity.IGetEditValue, Serenity.ISetEditValue {

        protected getIdProperty() { return "__id"; }

        private nextId = 1;

        constructor(container: JQuery) {
            super(container);
        }

        protected id(entity: TEntity) {
            return (entity as any)[this.getIdProperty()];
        }

        protected save(opt: Serenity.ServiceOptions<any>, callback: (r: Serenity.ServiceResponse) => void) {
            var request = opt.request as Serenity.SaveRequest<TEntity>;
            var row = Q.deepClone(request.Entity);

            var id = this.id(row);
            if (id == null) {
                (row as any)[this.getIdProperty()] = "`" + this.nextId++;
            }

            if (!this.validateEntity(row, id)) {
                return;
            }

            var items = this.view.getItems().slice();
            if (id == null) {
                items.push(row);
            }
            else {
                var index = Q.indexOf(items, x => this.id(x) === id);
                items[index] = Q.deepClone({} as TEntity, items[index], row);
            }

            this.setEntities(items);
            callback({});
        }

        protected deleteEntity(id: number) {
            this.view.deleteItem(id);
            return true;
        }

        protected validateEntity(row: TEntity, id: number) {
            return true;
        }

        protected setEntities(items: TEntity[]) {
            this.view.setItems(items, true);
        }

        protected getNewEntity(): TEntity {
            return {} as TEntity;
        }

        protected getButtons(): Serenity.ToolButton[] {
            return [{
                title: this.getAddButtonCaption(),
                cssClass: 'add-button',
                onClick: () => {
                    this.createEntityDialog(this.getItemType(), dlg => {
                        var dialog = dlg as GridEditorDialog<TEntity>;
                        dialog.onSave = (opt, callback) => this.save(opt, callback);
                        dialog.loadEntityAndOpenDialog(this.getNewEntity());
                    });
                }
            }];
        }

        protected editItem(entityOrId: any): void {

            var id = entityOrId;
            var item = this.view.getItemById(id);
            this.createEntityDialog(this.getItemType(), dlg => {
                var dialog = dlg as GridEditorDialog<TEntity>;
                dialog.onDelete = (opt, callback) => {
                    if (!this.deleteEntity(id)) {
                        return;
                    }
                    callback({});
                };

                dialog.onSave = (opt, callback) => this.save(opt, callback);
                dialog.loadEntityAndOpenDialog(item);
            });;
        }

        public getEditValue(property, target) {
            target[property.name] = this.value;
        }

        public setEditValue(source, property) {
            this.value = source[property.name];
        }

        public get value(): TEntity[] {
            var p = this.getIdProperty();
            return this.view.getItems().map(x => {
                var y = Q.deepClone(x);
                var id = y[p];
                if (id && id.toString().charAt(0) == '`')
                    delete y[p];
                return y;
            });
        }

        public set value(value: TEntity[]) {
            var p = this.getIdProperty();
            this.view.setItems((value || []).map(x => {
                var y = Q.deepClone(x);
                if ((y as any)[p] == null)
                    (y as any)[p] = "`" + this.nextId++;
                return y;
            }), true);
        }

        protected getGridCanLoad() {
            return false;
        }

        protected usePager() {
            return false;
        }

        protected getInitialTitle() {
            return null;
        }

        protected createQuickSearchInput() {
        }
    }
}